src/main/java/io/gravitee/policy/trafficshadowing/TrafficShadowingPolicy.java

```diff
@@ -57,65 +59,94 @@ public ReadWriteStream<Buffer> onRequestContent(ExecutionContext context) {
+            return new ShadowingReadWriteStream(proxyRequest, context, endpoint);
         }
 
         return null;
     }
 
     private HttpHeaders addShadowingHeaders(HttpHeaders headers, ExecutionContext context) {
-        HttpHeaders httpHeaders = new HttpHeaders(headers);
+        HttpHeaders httpHeaders = HttpHeaders.create(headers);
         if (configuration.getHeaders() != null) {
             configuration
                 .getHeaders()
-                .forEach(
-                    header -> {
-                        if (header.getName() != null && !header.getName().trim().isEmpty()) {
-                            try {
-                                String extValue = (header.getValue() != null)
-                                    ? context.getTemplateEngine().convert(header.getValue())
-                                    : null;
-                                if (extValue != null) {
-                                    httpHeaders.set(header.getName(), extValue);
-                                }
-                            } catch (Exception ex) {
-                                // Do nothing
-                                ex.printStackTrace();
+                .forEach(header -> {
+                    if (header.getName() != null && !header.getName().trim().isEmpty()) {
+                        try {
+                            String extValue = (header.getValue() != null) ? context.getTemplateEngine().convert(header.getValue()) : null;
+                            if (extValue != null) {
+                                httpHeaders.set(header.getName(), extValue);
                             }
+                        } catch (Exception ex) {
+                            // Do nothing
+                            ex.printStackTrace();
                         }
                     }
-                );
+                });
         }
         return httpHeaders;
     }
+
+    private static class ShadowingReadWriteStream extends SimpleReadWriteStream<Buffer> {
+
+        ProxyConnection connection;
+
+        List<Buffer> chunkBuffers = new ArrayList<>();
+
+        public ShadowingReadWriteStream(ProxyRequest proxyRequest, ExecutionContext context, ProxyEndpoint endpoint) {
+            endpoint
+                .connector()
+                .request(
+                    proxyRequest,
+                    context,
+                    connection -> {
+                        this.connection = connection;
+
+                        connection.responseHandler(response -> {
+                            LOGGER.debug("Traffic shadowing status is: {}", response.status());
+
+                            response.bodyHandler(buffer -> {}).endHandler(handler -> {});
+                        });
+                    }
+                );
+        }
+
+        @Override
+        public SimpleReadWriteStream<Buffer> write(Buffer chunk) {
+            if (connection == null) {
+                chunkBuffers.add(chunk);
```

Comment:
We should calculate the total retained size here (ie sum buffer.length) and if it exceed a certain size, just give up to preserve memory (maybe log a warn)